File: felicia/examples/learn/message_communication/ros_msg/python/ros_msg_subscribing_node.py
import felicia_py as fel
from felicia.core.protobuf.channel_pb2 import ChannelDef
from felicia.core.protobuf.master_data_pb2 import TopicInfo

from felicia.examples.learn.message_communication.ros_msg._SimpleMessage import SimpleMessage


class RosMsgSubscribingNode(fel.NodeLifecycle):
    def __init__(self, node_create_flag):
        super().__init__()
        self.topic = node_create_flag.topic_flag.value
        self.subscriber = fel.communication.Subscriber()
        self.subscriber.set_message_impl_type(TopicInfo.ROS)

    def on_did_create(self, node_info):
        self.node_info = node_info
        self.request_subscribe()

    def on_message(self, message):
        print("message : {}".format(message))

    def request_subscribe(self):
        settings = fel.communication.Settings()
        settings.buffer_size = fel.Bytes.from_bytes(512)

        self.subscriber.request_subscribe(self.node_info, self.topic,
                                          ChannelDef.CHANNEL_TYPE_TCP |
                                          ChannelDef.CHANNEL_TYPE_UDP |
                                          ChannelDef.CHANNEL_TYPE_UDS |
                                          ChannelDef.CHANNEL_TYPE_SHM,
                                          SimpleMessage, settings,
                                          self.on_message)

File: felicia/examples/learn/message_communication/ros_msg/cc/ros_msg_subscribing_node.cc
#include "felicia/examples/learn/message_communication/ros_msg/cc/ros_msg_subscribing_node.h"

#include "felicia/core/lib/strings/str_util.h"
#include "felicia/core/master/master_proxy.h"

namespace felicia {

RosMsgSubscribingNode::RosMsgSubscribingNode(
    const NodeCreateFlag& node_create_flag)
    : node_create_flag_(node_create_flag),
      topic_(node_create_flag_.topic_flag()->value()) {}

void RosMsgSubscribingNode::OnDidCreate(const NodeInfo& node_info) {
  node_info_ = node_info;
  RequestSubscribe();
}

void RosMsgSubscribingNode::OnMessage(SimpleMessage&& message) {
  std::cout << "message : " << message << std::endl;
}

void RosMsgSubscribingNode::RequestSubscribe() {
  communication::Settings settings;
  if (StartsWith(topic_, "ros://")) {
    settings.channel_settings.use_ros_channel = true;
  }
  settings.buffer_size = Bytes::FromBytes(512);

  subscriber_.RequestSubscribe(
      node_info_, topic_,
      ChannelDef::CHANNEL_TYPE_TCP | ChannelDef::CHANNEL_TYPE_UDP |
          ChannelDef::CHANNEL_TYPE_UDS | ChannelDef::CHANNEL_TYPE_SHM,
      settings,
      base::BindRepeating(&RosMsgSubscribingNode::OnMessage,
                          base::Unretained(this)));
}

}  // namespace felicia